test/api.spec.js

```diff
@@ -0,0 +1,325 @@
+      fetchNext: sandbox.stub().callsArgWith(0, resultSet, null, null, true),
+      getOptions: sandbox.stub().returns(queryOptions),
+      isComplete: sandbox.stub().returns(false),
+    };
+
+   api = proxyquire('../src/api', {
+      './helperClasses': {
+        MemcachedManager: sinon.stub().callsFake(() => {return cacheManager;}),
+        BigQueryManager: sinon.stub().callsFake(()=> {return bqManager;}),
+      },
+      'fs': {
+        readFileSync: sinon.stub().callsFake(() => { return 'fake-query-string';}),
+      },
+    });
+    app = express();
+    api(app);
+    request = supertest(app);
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+    sandbox.reset();
+  });
+  it('it should successfully stub the dependencies', () => {
+    cacheManager.createKey().should.equal(`__test__${token}${attrId}${from}`);
+    cacheManager.cacheResults.should.exist;
+    cacheManager.removeCache.should.exist;
+    cacheManager.get.should.exist;
+    cacheManager.set.should.exist;
+    bqManager.start.should.exist;
+    bqManager.fetchNext.should.exist;
+    bqManager.getOptions().should.deep.equal(queryOptions);
+    bqManager.isComplete().should.equal(false);
+  })
+  it('we can make a get request at the api endpoint', (done) => {
+    request
+        .get('/fetch_latest')
+        .query({
+          app_id: 'com.eduapp4syria.feedthemonsterENGLISH',
+          attribution_id: 'referral_source_8675309',
+          from: 0,
+        })
+        .expect('Content-Type', /json/)
+        .expect(200)
+        .expect((err, res)=> {
+          if(err) throw err;
+          console.log('res data: ');
+          console.log(res.data);
+          res.status.should.equal(200);
+        }).end(done);
+  });
+  it('we can successfully omit the referral id if desired', (done) => {
+    request
+        .get('/fetch_latest')
+        .query({
+          app_id: 'com.eduapp4syria.feedthemonsterENGLISH',
+          from: 0,
+        })
+        .expect(200)
+        .expect((err, res)=> {
+          if(err) throw err;
+          res.status.should.equal(200);
+        }).end(done);
+      });
+  it('we receive a 400 error if we omit the app id', (done) => {
+    request
+        .get('/fetch_latest')
+        .query({
+          attribution_id: 'referral_source_8675309',
+          from: 0,
+        })
+        .expect(400)
+        .expect((err, res)=> {
+          if (err) throw err;
+          res.status.should.equal(400);
+        }).end(done);
+  });
+  it('we receive a 400 error if we omit the cursor', (done) => {
+    request
+        .get('/fetch_latest')
+        .query({
+          app_id: 'com.eduapp4syria.feedthemonsterENGLISH',
+          attribution_id: 'referral_source_8675309',
+        })
+        .expect(400)
+        .expect((err, res)=> {
+          if (err) throw err;
+          res.status.should.equal(400);
+        }).end(done);
+  });
+  it('we receive a 500 error if we submit a missing ad id', (done) => {
```

Comment:
Should probably be 404, no?

Reply:
You're right that it's probably not a 500 error, but if the user provides a parameter that we don't have, is that a 404 File Not Found or a 200 with a "no data for this app id" response?

Reply:
> Status codes are a bit subjective! I like 404 for anything that is valid in theory but isnt found in the database. But if you like another 4** code better, go for it!

404 it is!